lib/action_dispatch/session/active_record_store.rb

```diff
@@ -117,30 +117,46 @@ def delete_session(request, session_id, options)
 
       def get_session_model(request, id)
         logger.silence_logger do
-          model = @@session_class.find_by_session_id(id)
-          if !model
+          model = get_session_with_fallback(id)
+          unless model
             id = generate_sid
-            model = @@session_class.new(:session_id => id, :data => {})
+            model = @@session_class.new(:session_id => id.private_id, :data => {})
             model.save
           end
           if request.env[ENV_SESSION_OPTIONS_KEY][:id].nil?
             request.env[SESSION_RECORD_KEY] = model
           else
             request.env[SESSION_RECORD_KEY] ||= model
           end
-          model
+          [model, id]
+        end
+      end
+
+      def get_session_with_fallback(sid)
+        if sid && !self.class.private_session_id?(sid.public_id)
```

Comment:
Why this check? How the user will get access to a private id?

Reply:
If we don't add this protection, the fallback mechanism allows an attacker to guess the private id  with a timing attack similar to the original vulnerability.

Reply:
In the hope to make things clearer and speed the whole process up a bit:

Imagine `public_id` just being a `private_id` in disguise. The attacker would send an `public_key` which is in fact a `private_id`! Line 136 will fail because `public_id` (aka "evil private_id") will be a double hashed `private_id`. This will cause line 139 to be executed. Which is basically the same insecure behaviour is as it is right now.
This line prevents this. If `public_id` is in fact a `private_id` it won't get used at all and no lookup will be performed.
Instead a new key/session will be generated and used ✅